### ISSUE TITLE
Fix "Rest element must be last element" error for tuples

### ISSUE THREAD

**Issue by LeoMosiah** [stylelint/stylelint#5195]
> Clearly describe the bug

There are false positives for two `Rest element must be last` when using rest elements in tuple types. Typescript version 4.2.0 added the feature to allow rest elements to occur anywhere within a tuple, trying to use that feature is giving a false positive on stylelint. 

> Which rule, if any, is the bug related to?

`Rest element must be last element`

> What code is needed to reproduce the bug? 

```ts

type Object = {
  foo: string
}

type Object2 = {
  bar: string
}

type TupleType = [...Object[], Object2]

```

> What stylelint configuration is needed to reproduce the bug?

```json
{
  "rules": {
    "number-leading-zero": "always"
  }
}
```

> Which version of stylelint are you using?

`13.11.0`

> How are you running stylelint: CLI, PostCSS plugin, Node.js API?

"CLI with `stylelint './src/**/*.{ts,tsx}`"

> Does the bug relate to non-standard syntax (e.g. SCSS, Less etc.)?

"Yes, it's related to TypeScript syntax."

> What did you expect to happen?

"No warnings to be flagged."

> What actually happened (e.g. what warnings or errors did you get)?

"The following warnings were flagged:"

```shell
index.tsx
21:55    Rest element must be last element
```

**Comment by jeddy3** [comment]
Thanks for the report and for using the template.

The error [appears to be coming from babel](https://github.com/babel/babel/blob/b3e2bcda73dea7d68b4c82bfabb92acb11b1ed90/packages/babel-parser/src/parser/error-message.js#L53).

It's likely a problem with [the parser](https://github.com/stylelint/postcss-css-in-js) stylelint uses [to support CSS-in-JS](https://stylelint.io/about/syntaxes).

It may simply need to use the latest version of babel. The package was [recently updated](https://github.com/stylelint/postcss-css-in-js/commit/5a2521d99ac4a756cbe723efcaf86321be862f0e), but not released. It'd be great to add a test to the package for this new syntax. If it passes, we can release a new version of the syntax.

Please consider [contributing to the parser](https://stylelint.io/developer-guide/syntaxes) if you have time.

**Comment by jeddy3** [comment]
Closing as the [`syntax` option was removed](https://github.com/stylelint/stylelint/issues/5289) for the [14.0.0 release](https://github.com/stylelint/stylelint/issues/5205). (See the [migration guide](https://github.com/stylelint/stylelint/blob/main/docs/migration-guide/to-14.md#syntax-option-and-automatic-inferral-of-syntax).)

We'll also [be deprecating](https://github.com/stylelint/postcss-css-in-js/issues/225) our [forked CSS-in-JS package](https://github.com/stylelint/postcss-css-in-js), in favour of leaner custom syntaxes.

Please consider [writing a custom syntax](https://github.com/stylelint/stylelint/blob/main/docs/developer-guide/syntaxes.md) specific to the CSS-in-JS library in this issue and fixing the bug there.
